core/lib/dal/src/consensus_dal.rs

```diff
@@ -30,54 +30,125 @@ impl ConsensusDal<'_, '_> {
         else {
             return Ok(None);
         };
-        Ok(Some(zksync_protobuf::serde::deserialize(row.state)?))
+        let Some(genesis) = row.genesis else {
+            return Ok(None);
+        };
+        Ok(Some(zksync_protobuf::serde::deserialize(genesis)?))
     }
 
-    /// Sets the current BFT replica state.
-    pub async fn set_replica_state(&mut self, state: &ReplicaState) -> sqlx::Result<()> {
-        let state =
-            zksync_protobuf::serde::serialize(state, serde_json::value::Serializer).unwrap();
+    /// Attempts to update the genesis.
+    /// Fails if the storage contains a newer genesis (higher fork number).
+    /// Noop if the new genesis is the same as the current one.
+    /// Resets the stored consensus state otherwise and purges all certificates.
+    pub async fn try_update_genesis(&mut self, genesis: &validator::Genesis) -> anyhow::Result<()> {
+        let mut txn = self.storage.start_transaction().await?;
+        if let Some(got) = txn.consensus_dal().genesis().await? {
+            // Exit if the genesis didn't change.
+            if &got == genesis {
+                return Ok(());
+            }
+            if got.fork.number >= genesis.fork.number {
+                anyhow::bail!("transition to a past fork is not allowed");
+            }
```

Comment:
Nit: You can use `anyhow::ensure` for these checks. Also, it could make sense to include details (`fork.number`?) into the error message for debug purposes.

Reply:
done